### ISSUE TITLE
Misuse of aria-label flags some correct uses

### ISSUE THREAD

**Issue by dd8** [validator/validator#775]
The following are reported as aria-label misuse, but the aria-label is read out by AT, so I don't think these are misuse:

```
<img aria-label="test" src="bar" />
```

```
<map name="infographic">
    <area shape="rect" coords="184,6,253,27"
          href="https://mozilla.org"
          target="_blank" aria-label="test" />
```

Other elements it warns on, but may not be misuse:
- canvas (using aria-label as ALT text)
- figure/figcaption (does aria-label replace figcaption text?)
- fieldset/legend (does aria-label replace legend text?)
- option (does aria-label replace option text?)

These are not reported, but adding aria-label means `role=presentation` is ignored:
https://www.w3.org/TR/wai-aria-practices-1.1/#presentation_role_ignored

```
<div aria-label="test" role="presentation" ></div>
<div aria-label="test" role="none" ></div>
```

**Comment by sideshowbarker** [comment]
> ```
> <img aria-label="test" src="bar" />
> ```

Thanks, agreed. `img` with `aria-label` is definitely not misuse per the discussion at https://github.com/w3c/html-aam/issues/160 and per https://developer.paciellogroup.com/blog/2017/07/short-note-on-aria-label-aria-labelledby-and-aria-describedby/ and https://www.w3.org/TR/using-aria/#practical-support-aria-label-aria-labelledby-and-aria-describedby

So I will fix that soon.

> ```
> <map name="infographic">
>     <area shape="rect" coords="184,6,253,27"
>           href="https://mozilla.org"
>           target="_blank" aria-label="test" />
> ```

`area` isn’t mentioned at all in https://developer.paciellogroup.com/blog/2017/07/short-note-on-aria-label-aria-labelledby-and-aria-describedby/ or https://www.w3.org/TR/using-aria/#practical-support-aria-label-aria-labelledby-and-aria-describedby, so I’m not sure what to do for that case.

To get clarification, maybe you could comment at https://github.com/w3c/html-aam/issues/160?

> Other elements it warns on, but may not be misuse:
> 
> * canvas (using aria-label as ALT text)
> * figure/figcaption (does aria-label replace figcaption text?)
> * fieldset/legend (does aria-label replace legend text?)

Warning for those aligns with https://developer.paciellogroup.com/blog/2017/07/short-note-on-aria-label-aria-labelledby-and-aria-describedby/ or https://www.w3.org/TR/using-aria/#practical-support-aria-label-aria-labelledby-and-aria-describedby as far as I can see. But if that guidance seems incomplete, please consider commenting at https://github.com/w3c/html-aam/issues/160

> * option (does aria-label replace option text?)

Warning for `option` aligns with the guidance as far as I understand it. But maybe I’m misreading.

> These are not reported, but adding aria-label means `role=presentation` is ignored
> https://www.w3.org/TR/wai-aria-practices-1.1/#presentation_role_ignored
> 
> ```
> <div aria-label="test" role="presentation" ></div>
> <div aria-label="test" role="none" ></div>
> ```

OK yeah I guess we should add an exception for the cases of `div`, etc., with role=presentation and role=none cases, and warn for those too.

**Comment by dd8** [comment]
`aria-label=` is pretty well supported on `area` elements - support is almost identical to `area alt`
https://www.powermapper.com/tests/screen-readers/labelling/area-aria-label/
https://www.powermapper.com/tests/screen-readers/labelling/area-alt/

I guess there are two facets to this:

a) using aria-label in places it doesn't make sense (https://github.com/w3c/html-aam/issues/160)
b) using aria-label where there's patchy support 

Warning about a) makes sense, since it's a misuse. 

Warning about b) doesn't seem like a a good idea since no a11y feature has perfect support. For example `alt` is very well supported on `img` but less well supported on `area` (although it was added to HTML in the 1990s).

Disclosure: Our product does check for a11y compatibility but this feature is turned off by default since there are some places you can't get a working solution on all screen readers

I'll run some tests and comment on https://github.com/w3c/html-aam/issues/160

**Comment by dd8** [comment]
This part of the AAM mappings specifies which elements aria-label is should produce a result
https://www.w3.org/TR/html-aam-1.0/#accessible-name-and-description-computation

JAWS 2018 / Firefox 60 aria-label support is a good example of why I think why a misuse warning shouldn't be tied to actual support. Using aria-label on `a` elements worked in previous versions of JAWS/Firefox but no longer work in JAWS 2018 with Firefox 60, although it still work in JAWS 2018 with IE11 (likely due to new Firefox Quantum engine) 

https://www.powermapper.com/tests/screen-readers/labelling/a-aria-label/

**Comment by dd8** [comment]
html-aam has been updated with an accessible name computation for `area` elements which uses aria-label

https://github.com/w3c/html-aam/issues/176
https://w3c.github.io/html-aam/#area-element

**Comment by dd8** [comment]
Other elements which specify aria-label as part of the name calculation but VNU flags as misuse:

`fieldset` https://w3c.github.io/html-aam/#fieldset-element-accessible-name-computation
`summary` https://w3c.github.io/html-aam/#summary-element-accessible-name-computation
`figure` https://w3c.github.io/html-aam/#figure-and-figcaption-elements

**Comment by dd8** [comment]
I've gone through the https://www.w3.org/TR/html-aam-1.0/ spec to produce a set of test cases for elements where html-aam specifies `aria-label` as part of the name calculation. Test code here:

```
<html>
<body aria-label="test">
    <!-- 5.1.1. -->
    <input aria-label='test' />
    <input aria-label='test' type="text"/>
    <input aria-label='test' type="password"/>
    <input aria-label='test' type="search"/>
    <input aria-label='test' type="tel"/>
    <input aria-label='test' type="url"/>
    <textarea aria-label='test' cols="1" rows="1"></textarea>
    <!-- 5.2.1. -->
    <input aria-label='test' type="button" value="Push me"/>
    <input aria-label='test' type="submit"/>
    <input aria-label='test' type="reset"/>
    <!-- 5.3.1 -->
    <input aria-label='test' type="image" alt="test"/>
    <!-- 5.4.1 -->
    <button aria-label='test'></button>
    <!-- 5.5.1 -->
    <fieldset aria-label='test'></fieldset>
    <!-- 5.6.1 -->
    <output aria-label='test'></output>
    <!-- 5.7.1 other form elements -->
    <input aria-label='test' type="checkbox"/>
    <input aria-label='test' type="color"/>
    <input aria-label='test' type="date"/>
    <input aria-label='test' type="datetime"/>
    <input aria-label='test' type="datetime-local"/>
    <input aria-label='test' type="email"/>
    <input aria-label='test' type="file"/>
    <input aria-label='test' type="month"/>
    <input aria-label='test' type="number"/>
    <input aria-label='test' type="radio"/>
    <input aria-label='test' type="range"/>
    <input aria-label='test' type="time"/>
    <input aria-label='test' type="week"/>
    <keygen aria-label='test'/>                    
    <select aria-label='test'><optgroup label="Test"><option>test</option></optgroup></select>
    <meter aria-label='test' value='0'></meter>
    <progress aria-label='test'></progress>
    <!-- 5.8.1 -->
    <details> <summary aria-label='test'>I have keys but no doors. I have space but no room. You can enter but can’t leave. What am I?</summary>A keyboard.</details>
    <!-- 5.9.1 -->
    <figure aria-label='test'><img src="elephant.jpg" alt="Elephant at sunset"/><figcaption>An elephant at sunset</figcaption></figure>
    <!-- 5.10.1 -->
    <img aria-label='test' alt="test" src="test.png" />
    <!-- 5.11.1 -->
    <table aria-label='test'><tbody><tr><th>Header</th><td>Data</td></tr></tbody></table>
    <!-- 5.12.1 -->
    <table><tbody><tr aria-label='test'><th aria-label='test'>Header</th><td aria-label='test'>Data</td></tr></tbody></table>	
    <!-- 5.13.1 -->
    <a aria-label='test' href="/">Click Here</a>	
    <!-- 5.14.1. -->
    <map name="test"><area aria-label="test" alt="test" href="/"/></map>
    <!-- 5.15.1 -->
    <iframe aria-label="test" title="test" src="/"></iframe>
    <!-- 5.16.1 Section and Grouping elements -->
    <!-- https://html.spec.whatwg.org/multipage/sections.html#sections - see above for body -->
    <article aria-label="test" ></article>
    <section aria-label="test" ></section>
    <nav aria-label="test" ></nav>
    <aside aria-label="test" ></aside>
    <!-- https://html.spec.whatwg.org/multipage/grouping-content.html#grouping-content -->
    <p aria-label="test"></p>
    <hr aria-label="test" />
    <pre aria-label="test"></pre>
    <blockquote aria-label="test"></blockquote>
    <ol aria-label="test"><li></li></ol>
    <ul aria-label="test"><li></li></ul>
    <menu aria-label="test"><li></li></menu>
    <ul><li aria-label="test"></li></ul>
    <dl aria-label="test"><dt>What do we want</dt><dd>More tests</dd></dl>
    <dl><dt aria-label="test">What do we want</dt><dd>More tests</dd></dl>
    <dl><dt>What do we want</dt><dd aria-label="test">More tests</dd></dl>
    <main aria-label="test"></main>
    <div aria-label="test"></div>
    <!-- 5.17.1 text elements not listed elsewhere -->
    <abbr aria-label="test"></abbr>
    <b aria-label="test"></b>
    <bdi aria-label="test"></bdi>
    <bdo aria-label="test" dir="rtl"></bdo>
    <br aria-label="test"/>
    <cite aria-label="test"></cite>
    <code aria-label="test"></code>
    <dfn aria-label="test"></dfn>
    <em aria-label="test"></em>
    <i aria-label="test"></i>
    <kbd aria-label="test"></kbd>
    <mark aria-label="test"></mark>
    <q aria-label="test"></q>
    <ruby aria-label="test">字 <rp>(</rp><rt>ji</rt><rp>)</rp></ruby>
    <ruby>字 <rp aria-label="test">(</rp><rt>ji</rt><rp>)</rp></ruby>
    <ruby>字 <rp>(</rp><rt aria-label="test">ji</rt><rp>)</rp></ruby>
    <s aria-label="test"></s>
    <samp aria-label="test"></samp>
    <small aria-label="test"></small>
    <strong aria-label="test"></strong>
    <sub aria-label="test"></sub>
    <sup aria-label="test"></sup>
    <time aria-label="test">2019-07-07</time>
    <u aria-label="test"></u>
    <var aria-label="test"></var>
    <wbr aria-label="test"/>					
</body>
</html>
```

Feel free to use these tests in WPT or anywhere else.

A few points worth noting:

- there's a proposal to prohibit aria-label on `div` and `span` https://github.com/w3c/html-aam/issues/160#issuecomment-499959873
- the spec allows it but does` aria-label` makes sense on `ruby` `rp` and `rt` ?
- `aria-label` is currently allowed on `body` (because it's a sectioning element) I'm not sure this makes sense so going to report as an html-aam issue.

**Comment by dd8** [comment]
WAI-ARIA Authoring Practices 1.1 published on Aug 14 has detailed per-role guidance on using aria-label here:

https://www.w3.org/TR/wai-aria-practices-1.1/#naming_with_aria-label
https://www.w3.org/TR/wai-aria-practices-1.1/#naming_role_guidance

There's at least one case (from https://lists.w3.org/Archives/Public/www-validator/2019Sep/0007.html) where aria-label is allowed by wai-aria-practices (and described as potentially beneficial) but VNU is flagging misuse:

```
<ul  aria-label="test"><li></li></ul> <!-- role=list naming: Discretionary -->
```

but this should still flag misuse:

```
<ul><li aria-label="test"></li></ul> <!-- role=listitem naming: Do not name -->
```

**Comment by sideshowbarker** [comment]
Closing per recent updates.

**Comment by sideshowbarker** [comment]
> `<table><tbody><tr aria-label='test'><th aria-label='test'>Header</th><td aria-label='test'>Data</td></tr></tbody></table>`

For td/th/tr, https://w3c.github.io/html-aam/#other-tabular-data-elements says:

> If the table element has an aria-label or an aria-labelledby attribute the accessible name is to be calculated using the algorithm defined in Accessible Name and Description: Computation and API Mappings 1.1.

That is, it explicitly says “If the table element” — not just “If the element”.

So doesn’t that mean the requirement is to essentially ignore whatever `aria-label` attribute a `td`, `th`, or `tr` element might have, and instead look back up the tree and use whatever `aria-label` attribute the ancestor `table` element has?

In other words, as far as the checker behavior goes, doesn’t that mean it’s appropriate for the checker to warn about not using `aria-label` with `td`, `th`, or `tr` elements?

**Comment by sideshowbarker** [comment]
Update: See https://github.com/w3c/html-aam/issues/335#issuecomment-851476428

> That said, while using `aria-label` on these elements is valid and is meant to be respected when determining the name of the elements by browsers, as Steve noted in his html5accessibility post, support for `aria-label` with AT is rather spotty.
> 
> For instance [with this table](https://encoding.spec.whatwg.org/windows-1252.html) NVDA with Firefox and Chromium ignore the `aria-label` when navigating the list.  So none of the coloring of the table will be accessible for people using this browser/AT combo.
> 
> With VoiceOver, on macOS (but not iOS) the `aria-label` is respected but that makes all cells initially announce as the `aria-label` value (e.g., "Two bytes"), requiring a user to navigate into the cell to determine what the actual value presented is.
> 
> tldr; these tables are valid, but the legend information is either not going to be announced, or it is but it will create an extra step for AT users to hear the actual content of the table cell.

So the bottom line is that since current UAs/AT don’t hanadle `aria-label` with `td`, `th`, and `tr` as expected, it’s appropriate for the checker to warn about those cases.

**Comment by dd8** [comment]
I would draw a distinct line between 2 cases:

a) `aria-label` is specified as not allowed/ignored when exposing an element's content to AT 
b) `aria-label` is specified as allowed/used on an element but has poor support 

Case a) is misuse, but warning that case b) is "misuse" may create a a feedback loop with AT implementors who don't implement the specified behaviour because VNU flags it as misuse. 

If case b) warns that about poor support instead, that's much less of a risk (and may even put some pressure on implementors to implement the spec).

**Comment by sideshowbarker** [comment]
> If case b) warns that about poor support instead, that's much less of a risk (and may even put some pressure on implementors to implement the spec).

Agreed. So I’ll do an update to the code which makes a distinction for those cases, and for case _b_, emits an appropriate *“not yet well supported”* (or whatever the optimal wordig is) message. (Suggestions for the specific wording very much welcome.)
